### PR TITLE
[LSM] Rebalance LSM Delegations

### PR DESCRIPTION
**Pull request by sampocs**
## Terminology
**UnbalancedDelegation**: represents stake from LSM tokens that have not yet been rebalanced according to our validator distribution.
**BalancedDistribution**: represents all other stake. This can either come from normal delegations or from LSM tokens that have already been rebalanced.

## Context
With LSM, we'll now have to rebalance stake from LSM tokens periodically. This is designated on the HostZone as the "Unbalanced Delegation". We can utilize the original rebalance function for this but with a slight parameter change that designates whether we want to rebalance the "Unbalanced" or "Balanced" portion. 
* _Rebalancing the Balanced Portion_: This is still be exposed as a transaction and can be called after validator weights change
* _Rebalancing the Unbalanced Portion_: This will occur epochly 

## Brief Changelog
* Fixed bugs in rebalancing
* Split rebalancing into sub functions
* Added parameter to rebalancing to designate "Unbalanced" vs "Balanced" delegations
* Modified rebalance ICA callback to handle the Unbalanced portion
* Added rebalancing unit tests
* Cleaned up main rebalancing function
* Made sorting of validators deterministic by checking the validator name if there's a tie on the weight 

**I'd recommend reviewing this commit by commit. The rebalancing code was moved to a new file so the GH diffs will show everything as net new.** 
Also it's a bit deceiving, but ([this commit](https://github.com/Stride-Labs/stride/pull/726/commits/9e69d8099a7990e14ae28732b839b4f29af11eaa)) only moved the function to a new file - it did not change any of the actual code.

### CODE REVIEW

File: x/stakeibc/keeper/validator_selection.go
Comment (riley-stride) [inline x/stakeibc/keeper/validator_selection.go:93]: Surplus/Deficit feels more accurate than Overweight/Underweight, because the * amount * is in surplus/deficit relative to the weight. The weight it always correct, so overweight/underweight is technically not correct

**Review by riley-stride — APPROVED** [review]
Looks good to me. Solid test coverage. Though I think we need a thorough review from @shellvish on the rebalancing portion.

File: x/stakeibc/keeper/validator_selection.go
Comment (ethan-stride) [reply x/stakeibc/keeper/validator_selection.go:93]: Agree

File: proto/stride/stakeibc/host_zone.proto
Comment (sampocs) [reply proto/stride/stakeibc/host_zone.proto:15]: ah yes, not sure how that got in there. removed

File: x/stakeibc/keeper/validator_selection_test.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection_test.go:189]: done

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go]: I think this is outdated and was resolved in a later commit

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go]: yeah same as above, this was changed in a later commit. I was trying to iterate on this function in chunks and didn't want to change too much at once

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go]: same answer as above

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go]: i believe this was also resolved in a later commit

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go]: also addressed in a later commit, glad we're on the same page though haha

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go]: also resolved in a later commit

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go]: this was a bug in the original implementation but was resolved in one of the later commits

File: x/stakeibc/keeper/icacallbacks.go
Comment (sampocs) [reply x/stakeibc/keeper/icacallbacks.go:13]: unbond/undelegate are synonymous. I'd rather not change these callback IDs unless we really need to cause it would require a store migration

File: x/stakeibc/keeper/lsm.go
Comment (sampocs) [reply x/stakeibc/keeper/lsm.go:224]: correct, just clarified the comment

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go]: yeah good point, let's table it for now though and revisit

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go:136]: The validatorDeltas array shouldn't contain any validator's with zero amount - they're filtered out [here](https://github.com/Stride-Labs/stride/pull/726/files#diff-7eb0c0f79f6e24ba948f073cca1f9f6848f614e735f052b4714b82657237ecebR198-R206)

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go:110]: agreed as will. Will update if @shellvish's on the same page!

File: x/stakeibc/keeper/validator_selection.go
Comment (shellvish) [inline x/stakeibc/keeper/validator_selection.go:119]: imo it's a little bit unintuitive that Surplus Validators have a negative delta. Should we invert the deltas, so define Surplus Validator as having positive deltas, and Deficit Validators as having negative deltas?

File: x/stakeibc/keeper/validator_selection.go
Comment (shellvish) [inline x/stakeibc/keeper/validator_selection.go:215]: great addition!

File: x/stakeibc/keeper/validator_selection.go
Comment (shellvish) [inline x/stakeibc/keeper/validator_selection.go:244]: I think so, because the list is actually ordered now! We should test it in integration tests though (I think because we're testing for stochasticity, it might be challenging to test through unit tests? low confidence). I _think_ the only integration test change would be adding a third validator and then setting two of them to the same weight, but please double-check me on that

Honestly what a silly piece of code, reverse the list and then stable sort 😂

File: x/stakeibc/keeper/validator_selection_test.go
Comment (shellvish) [inline x/stakeibc/keeper/validator_selection_test.go:216]: good edge case to cover!

File: x/stakeibc/keeper/validator_selection_test.go
Comment (shellvish) [inline x/stakeibc/keeper/validator_selection_test.go:110]: nit, can we make `val3` and `val4` -8 and -8? Just to add one more edge-case

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go:119]: yeah I thought the same thing! Will do

Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go:119]: One other option is to just have them all be positive and add a field to the Delta struct that says whether it was a surplus or deficit.

Not sure if this would clean it up or just make it more congested, but would be nice to not have to think about the sign at all!

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go:161]: hmm good question. I think it's still possible to hit this case but not be done with the algorithm

For instance, lets say you have
```
-10
-5
+5
+10
```
I think we see that the -10 matches with the +10, but if we didn't increment the indicies, we wouldn't get those 5'ers

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go:82]: hahahah we just love python so much

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go:244]: yeah good call, we should definitely be pretty thorough in IG testing the rebalancing before we release this! Added a TODO in the spec

File: x/stakeibc/keeper/validator_selection_test.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection_test.go:110]: I want to this in there cause this is the edge case from our earlier comments where all the delta from one val gets moved to the other val and they both increment indicies.

Is the case you wanna add one where two validator's have the same stake? Cause I can add that case in lower in this test!

Comment (sampocs) [reply x/stakeibc/keeper/validator_selection_test.go:110]: added another case, lmk if it resolves it for you

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go:124]: I'm definitely on board with adding redundant checks if they help readability! But just to make sure I'm following, you're thinking something like:
```go
if overWeightIndex == underWeightIndex {
  break
}
// or 
if overWeightValidator.Delta < underWeightValidator.Delta {
  break
}
```

Considering that that first case doesn't happen every time, and the second case assumes a knowledge of the delta signs (which is a bit hard to remember), I personally feel like they both make the code a little harder to follow, but will definitely defer to you since you're coming from the reviewer/reader perspective atm!

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go]: nvm, I think swapping the sign cleaned it up plenty

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go]: done

File: x/stakeibc/keeper/validator_selection.go
Comment (shellvish) [reply x/stakeibc/keeper/validator_selection.go]: Yeah I'm in favor of swapping signs for now! Not opposed to adding a third field too if you think it's still confusing though, esp given the density of this code

File: x/stakeibc/keeper/validator_selection.go
Comment (shellvish) [reply x/stakeibc/keeper/validator_selection.go]: Yes definitely, great catch! We absolutely should still update them

File: x/stakeibc/keeper/validator_selection.go
Comment (shellvish) [reply x/stakeibc/keeper/validator_selection.go]: Oh I was thinking something slightly different, more like
`
if overWeightIndex <= underWeightIndex {
  break
}
`
I think we always want to break if the two indices have flipped!

File: x/stakeibc/keeper/validator_selection.go
Comment (shellvish) [reply x/stakeibc/keeper/validator_selection.go]: Nice, thank you!

File: x/stakeibc/keeper/validator_selection.go
Comment (shellvish) [reply x/stakeibc/keeper/validator_selection.go:82]: is that why they call it snake case?? 😮

File: x/stakeibc/keeper/validator_selection_test.go
Comment (shellvish) [reply x/stakeibc/keeper/validator_selection_test.go]: Ah gotcha that makes sense, yeah agreed that resolves it for me! Thanks for making that change!

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go]: just added!

File: x/stakeibc/keeper/validator_selection.go
Comment (sampocs) [reply x/stakeibc/keeper/validator_selection.go]: reading over the code again, I don't think the additional field would make it any clearer, but we can always come back to it! Marking this resovled

**Pull request merged**
